### PR TITLE
unit test: refactored bindist.py

### PR DESCRIPTION
**Pull request by alalazo**
Modifications:
- Make use of SpackCommand objects and deduplicate code wherever possible
- Moved cleaning of mirrors to fixtures
- Ensure mock configuration has a clear initialization order

### CODE REVIEW

**Comment by alalazo** [comment]
@gartung I refactored a few tests you have added in #15687 in a way that is functionally equivalent but more compact. One question I have is if there are relevant assertions we can perform during tests to verify invariants related to our business logic of creating binary caches. Right now most of the tests just execute a series of operations and don't verify any outcome, so they are not as useful as they could be to protect against regression.

**Review by gartung — DISMISSED** [review]

File: lib/spack/spack/test/bindist.py
Comment (scottwittenburg) [inline lib/spack/spack/test/bindist.py:43]: Just curious: what's the practical goal of reducing the scope from `session` to `module`?  Is this about limiting the lifetime of the mirror to just the tests in this file?  I ask because I sometimes run into issues where it seems there must be binaries in a mirror which is configured in some test I'm working on, but I didn't set it up myself and so it must be laying around from some other test.

File: lib/spack/spack/test/bindist.py
Comment (scottwittenburg) [inline lib/spack/spack/test/bindist.py:92]: Thanks for the comment here.  One additional thing that would be nice for me to understand is *why* it's useful or necessary to have another substitution of config scopes here.  I suspect others might not need it spelled out, but somehow I do ;-)

File: lib/spack/spack/test/bindist.py
Comment (scottwittenburg) [inline lib/spack/spack/test/bindist.py:410]: For my clearer understanding: Is it a best practice to inject fixtures using `usefixtures` rather than directly into the parameter list?  Or is there some other reason for this change?

**Review by scottwittenburg — DISMISSED** [review]
Thanks @alalazo 😄 It's probably obvious from my review questions, but there are nuances here that I'm missing, so my approval is more of a courtesy than an indicator that everything looks right.  To my inexpert eye however, it all looks fine.

Something else I'm curious about: What brought you here to clean this up?  Specifically, were you running into leftovers from this module that were interfering with other tests, or are you just on a cleaning mission?

**Comment by alalazo** [comment]
> What brought you here to clean this up? Specifically, were you running into leftovers from this module that were interfering with other tests, or are you just on a cleaning mission?

Long story short, while working on unrelated things I stumbled on #21491 A seemingly simple one line fix for that issue lead to lot of errors in the tests, starting from this module. When I read it I thought I could clean it up and make it a few lines shorter, regardless of the originating issue, hence the PR.

File: lib/spack/spack/test/bindist.py
Comment (alalazo) [reply lib/spack/spack/test/bindist.py:43]: > Is this about limiting the lifetime of the mirror to just the tests in this file?

Yes. Each fixture has [a scope](https://docs.pytest.org/en/stable/reference.html#pytest-fixture-api) that by default is "function". The scope defines the granularity of the lifetime for the fixture, i.e. when setup and cleanup of the fixtures are called. Fixtures that are "function" scoped run setup before each test function and cleanup after the test function is run, while the ones that are e.g. "session" scoped run setup once before the first test in the suite that uses them and cleanup and the end of the entire test suite. In this case the fixture is defined in this module, and visible only there, so it doesn't make sense to leave artifacts around when tests from other modules are running.

Comment (scottwittenburg) [reply lib/spack/spack/test/bindist.py:43]: That's sort what I was guessing, in a vague way.  Thanks for the explanation, sorry to slow you down 😉

File: lib/spack/spack/test/bindist.py
Comment (alalazo) [reply lib/spack/spack/test/bindist.py:92]: >  why it's useful or necessary to have another substitution of config scopes here

This I don't know :slightly_smiling_face:  

I just noticed that the `default_config` fixture was conflicting with another fixture being used, i.e. `install_mockery`. The conflict is somehow hidden, since one must also know that:
-  `install_mockery` depends on another fixture (called `config`) that swaps in a mock configuration used only in tests
-  since `install_mockery` and `default_config` don't depend on each other, the initialization order was undefined

Here I established a clear order of initialization by making `default_config` dependent on `install_mockery`. Maybe further simplifications may be done here, but I didn't want to change the semantics of the tests.

Comment (scottwittenburg) [reply lib/spack/spack/test/bindist.py:92]: > * since `install_mockery` and `default_config` don't depend on each other, the initialization order was undefined

Ok, this makes sense and clears up my misunderstanding of your PR description comment.

Thanks for the explanation!

File: lib/spack/spack/test/bindist.py
Comment (alalazo) [reply lib/spack/spack/test/bindist.py:410]: I don't think we ever enforced that, but I usually tend to write tests where I put in `usefixtures` fixtures that are there only for setting a state (i.e. their return value is not used in tests). It's just a personal style preference though.

File: lib/spack/spack/test/bindist.py
Comment (alalazo) [reply lib/spack/spack/test/bindist.py:43]: >  Thanks for the explanation, sorry to slow you down

Hey, no worries and no slowing down at all :+1:

File: lib/spack/spack/test/bindist.py
Comment (becker33) [inline lib/spack/spack/test/bindist.py]: Can we use the non-deprecated format strings here?

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/bindist.py
Comment (gartung) [reply lib/spack/spack/test/bindist.py:92]: I needed a default and a non-default config for the directory layout. The only way I could get it to work was the hackery I did.

**Comment by gartung** [comment]
> @gartung I refactored a few tests you have added in #15687 in a way that is functionally equivalent but more compact. One question I have is if there are relevant assertions we can perform during tests to verify invariants related to our business logic of creating binary caches. Right now most of the tests just execute a series of operations and don't verify any outcome, so they are not as useful as they could be to protect against regression.

The tests were added for coverage, i.e. syntax error or typos. I did not want to have to run a full build just to exercise these functions. 

P.S. Have you verified that the tests work on macOS? Getting both linux and macOS to work was a major pain in the ass.

**Comment by alalazo** [comment]
> The tests were added for coverage, i.e. syntax error or typos.

Sure, my point is that tests are more meaningful if they also check invariants or expected states after executing some portion of the code. That can of course be done in a following PR, but the rationale is that a test without assertions is more prone to false negatives (tests that pass while something is wrong). I was wondering if you have in mind checks that may be relevant and should be added in the future.

**Comment by alalazo** [comment]
> Have you verified that the tests work on macOS? Getting both linux and macOS to work was a major pain in the ass.

The tests pass for me locally on both Ubuntu 18.04 and BigSur. Let's wait for the outcome of the current MacOS tests in CI, but as of 806f494 they were both passing

**Comment by gartung** [comment]
Asserts could be added to check that the relocated rpaths are as expected using the output of patchelf or install_name_tool.

File: lib/spack/spack/test/bindist.py
Comment (alalazo) [reply lib/spack/spack/test/bindist.py]: @becker33 Should be done

**Comment by alalazo** [comment]
> Asserts could be added to check that the relocated rpaths are as expected using the output of patchelf or install_name_tool.

Thanks for the hints. I'll take that into account for a following PR.

**Comment by alalazo** [comment]
@becker33

File: lib/spack/spack/test/bindist.py
Comment (tgamblin) [reply lib/spack/spack/test/bindist.py:410]: I usually find it clearer to read them as arguments but like Max said, we do not have a clear policy on this.  I don't feel too strongly about it -- his argument is valid.

File: lib/spack/spack/test/bindist.py
Comment (tgamblin) [inline lib/spack/spack/test/bindist.py]: Can you split this into pieces?

```python
scheme = os.path.join(
    "${name}",
    "${version}",
    "${architecture}-${compiler.name}-${compiler.version}-${hash}"
)
```

**Review by tgamblin — CHANGES_REQUESTED** [review]
One minor request.  I do not think we should have any E501's in the code if it's implementable any other way.

File: lib/spack/spack/test/bindist.py
Comment (alalazo) [reply lib/spack/spack/test/bindist.py]: @tgamblin Done

**Review by tgamblin — APPROVED** [review]

**Review by becker33 — APPROVED** [review]

**Pull request merged**
